ID = unique Id of each customer                     
Age   = Age of the customer                  
Experience = experience of the customer             
Income   = salary of each customer               
ZIP Code  = unique code of customer              
Family    = Type of family              
CCAvg     = average value             
Education = education of the customer             
Mortgage  =  Mortgage            
Personal Loan = Personal Loan          
Securities Account = Securities Account  
CD Account  =CD Account             
Online   = Online          
CreditCard= CreditCard 

## Step-1.... Import Libraries

In [1]:
## Import Libraries
import os
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
%matplotlib inline

## Step-2... Load data

In [5]:
## Read "UniversalBank.csv" using pandas
unibank = pd.read_csv("C:/Data sceince  data/UniversalBank.csv")

## Step-3...Information about data

In [6]:
## Print the first 7 rows
unibank.head(7)

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   
5   6   37          13      29     92121       4    0.4          2       155   
6   7   53          27      72     91711       2    1.5          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1  
5              0                   0           0       1           0  
6              0                   0           0       1           0

In [7]:
unibank.shape  ### Size of the data set

(5000, 14)

In [8]:
## Check the datatype of each variable
unibank.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [9]:
## Statistical summery of data set
unibank.describe()

ID          Age   Experience       Income      ZIP Code  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93152.503000   
std    1443.520003    11.463166    11.467954    46.033729   2121.852197   
min       1.000000    23.000000    -3.000000     8.000000   9307.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937938     1.881000    56.498800       0.096000   
std       1.147663     1.747659     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities Account  CD Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

## Step-4...Data Pre-Processing

In [10]:
## Drop columns which are not significant
unibank.drop(["ID","ZIP Code"],axis=1,inplace=True)

In [11]:
## Convert Categorical Columns to Dummies
cat_cols = ["Family","Education","Personal Loan","Securities Account","CD Account","Online","CreditCard"]
unibank = pd.get_dummies(unibank,columns=cat_cols,drop_first=True,)

In [12]:
len(unibank.columns) ### Length of columns

15

In [13]:
## Split the data into X and y
X = unibank.copy().drop("Personal Loan_1",axis=1)
y = unibank["Personal Loan_1"]

In [14]:
X.shape ## size of the x

(5000, 14)

In [15]:
y.shape ## size of y

(5000,)

In [16]:
## Split the data into X_train, X_test, y_train, y_test with test_size = 0.20 using sklearn
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

In [17]:
## Print the shape of X_train, X_test, y_train, y_test
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(4000, 14)
(1000, 14)
(4000,)
(1000,)


In [18]:
## Scale the numeric attributes
## import the library
from sklearn.preprocessing import StandardScaler
## Instantiating the object
scaler = StandardScaler()
### fit on train data
scaler.fit(X_train.iloc[:,:5])
## transform on test data
X_train.iloc[:,:5] = scaler.transform(X_train.iloc[:,:5])
X_test.iloc[:,:5] = scaler.transform(X_test.iloc[:,:5])

C:\Users\K SREELATHA\anaconda3\lib\site-packages\pandas\core\indexing.py:1736: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value[:, i].tolist())
C:\Users\K SREELATHA\anaconda3\lib\site-packages\pandas\core\indexing.py:1736: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value[:, i].tolist())


## Step-4... Model Building

In [19]:
## Build a Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier

## Create an RFC object and print it to see the default arguments
rfc = RandomForestClassifier()
rfc

RandomForestClassifier()

In [20]:
## Create a new RFC object with name "rfc_50trees_oob" 
## This object should have 50 trees and should also give us oob score
## So use arguments n_estimators=50,oob_score=True
## n_estimators are the number of trees

rfc_50trees_oob = RandomForestClassifier(n_estimators=50,oob_score=True)
rfc_50trees_oob

RandomForestClassifier(n_estimators=50, oob_score=True)

In [21]:
## Fit the model rfc_50trees_oob on the train data (X_train,y_train)
rfc_50trees_oob.fit(X = X_train,y = y_train)

RandomForestClassifier(n_estimators=50, oob_score=True)

In [22]:
## Predict on test data and store it in the variable y_pred
y_pred = rfc_50trees_oob.predict(X_test)

In [23]:
## Evalutaion
from sklearn.metrics import accuracy_score,confusion_matrix

In [24]:
## Print accuracy_score using y_test and y_pred
accuracy_score(y_test,y_pred)

0.988

In [25]:
## Print Confusion Matrix using y_test and y_pred
confusion_matrix(y_test,y_pred)

array([[901,   0],
       [ 12,  87]], dtype=int64)

In [26]:
rfc_50trees_oob.oob_score_

0.98425

## Parameter Tuning

In [27]:
## Use Grid Search for parameter tuning

from sklearn.model_selection import GridSearchCV

## n_jobs = -1 uses all cores of processor
## max_features is the maximum number of attributes to select for each tree
rfc_grid = RandomForestClassifier(n_jobs=-1, max_features='sqrt')
 
# Use a grid over parameters of interest
## n_estimators is the number of trees in the forest
## max_depth is how deep each tree can be
## min_sample_leaf is the minimum samples required in each leaf node for the root node to split
## "A node will only be split if in each of it's leaf nodes there should be min_sample_leaf"

param_grid = { 
           "n_estimators" : [9, 18, 27, 36, 45, 54, 63],
           "max_depth" : [5, 10],
           "min_samples_leaf" : [2, 4]}
 
rfc_cv_grid = GridSearchCV(estimator = rfc_grid, param_grid = param_grid, cv = 10)

In [28]:
## Fit the grid search model
rfc_cv_grid.fit(X = X_train, y = y_train)

GridSearchCV(cv=10,
             estimator=RandomForestClassifier(max_features='sqrt', n_jobs=-1),
             param_grid={'max_depth': [5, 10], 'min_samples_leaf': [2, 4],
                         'n_estimators': [9, 18, 27, 36, 45, 54, 63]})

In [29]:
## Print best score and parameters
print(rfc_cv_grid.best_score_,rfc_cv_grid.best_params_)

0.9842500000000001 {'max_depth': 10, 'min_samples_leaf': 2, 'n_estimators': 45}


### Step-5... conclusion

Random forest classifier is applied on the given dataset. accuracy scroe is 0.988.hyper parametr tunning gives us the best parameters such as max_depth as 10,min_sample_leaf as 2 and N-estimators as 45. This will be the best model